### ISSUE TITLE
Request: Remove UI

### ISSUE THREAD

**Issue by ghost** [intellum/neeman#21]
It would be great if the UI was removed from the framework target all together. I need very different UI for my project. Would you be interested in taking the project in that direction?

**Comment by onato** [comment]
Yes, I have been meaning to look into a more Swifty API that doesn't rely on inheritance, but haven't had much chance to think it through.

**Comment by ghost** [comment]
Let me see if I can put something together. It probably won't be a pull request though.

Be right back.

**Comment by ghost** [comment]
[Hybrid.zip](https://github.com/intellum/neeman/files/2639255/Hybrid.zip)

**Comment by onato** [comment]
This looks great. Nice work! I'm happy to start moving forward with this. Let's set up a feature branch. I added you as a collaborator.

My ProcessPool singleton still has the old pre-Swift-3 convention of `sharedInstance` instead of `shared`.

`shouldShowNewScene` should receive a `WKNavigationAction`. This allows you to check if the navigation was, for example, a redirect or a click using the `navigationType`.

**Comment by ghost** [comment]
I'm glad you like the code. 

I like how you have built a swift to js / css bridge. It would be very interesting to see your project develop more on this.

I don't know very much about website development so I don't fully understand `AtDocumentStart` or end. I can't read js or css ! 

I have no idea what is going on here.

```css
document.documentElement.style.webkitTouchCallout='none';

var root = document.getElementsByTagName( 'html' )[0]
var existingClass = root.getAttribute( 'class' )
root.setAttribute( 'class', 'neeman-hybrid-app neeman-hybrid-app-version-${VERSION} ' + existingClass );

var styleElement = document.createElement('style');
root.appendChild(styleElement);
styleElement.textContent = '${CSS}';
```

This prevents a highlight happening on touch?

```css
canvas {
    -webkit-tap-highlight-color: transparent;
}
```

I understand that you are intercepting messages with this. You have taken the link from the button, which is cool. But I can't write javascript and I'm too lazy to learn a new language.

```javascript
 (function(){
  var buttons = document.getElementsByClassName("header-context-button");
  if(buttons.length > 0) {
  var button = buttons[0];
  webkit.messageHandlers.didGetBarButton.postMessage(button.href);
  }
  
  var metaTags=document.getElementsByTagName("meta");
  
  var username = "";
  for (var i = 0; i < metaTags.length; i++) {
  if (metaTags[i].getAttribute("name") == "octolytics-actor-login") {
  username = metaTags[i].getAttribute("content");
  break;
  }
  }
  webkit.messageHandlers.didGetUserName.postMessage(username);
  })();
```

🤯

So I'm not sure how I can help you because my understanding of web development is zero.

Regarding a feature branch: I think maybe you should consider removing some of the UI logic before adding another feature.

**Comment by onato** [comment]
`AtDocumentStart` happens before the WebView begins to load the HTML into memory. Code loaded there won't have access to the DOM until the page loads fully.

> I have no idea what is going on here.
This code adds a class attribute to the HTML tag so you can add app specific CSS later. It also inserts the contents of the CSS files into the DOM. 

`-webkit-tap-highlight-color` prevents a canvas element from being highlighted when you tap it. I think I added that to prevent a highlight around an interactive graph when you tap on it.

I would have to get the Hybrid stuff running on a separate branch, port my apps to it making sure everything I need is supported and once it's ready, merge it into the main branch. I have quite a bit of code in a private repo that builds on Neeman, so it might take me a while.

What's your plan for using this? I'd be keen to hear more about your use-case.

**Comment by ghost** [comment]
Thanks for the explanation.

I have a new client that wants to build a hybrid web / native App and I'm searching GitHub for clever ways to achieve that. Project starts in January.

I have built stable hybrid Apps before but it wasn't fun.

Sound like you will be busy. I will watch for changes to your repo.
